### PR TITLE
Irwaters/paas 194

### PR DESCRIPTION
**Pull request by irwaters**
gathers the status.hostIP from the pod status api, writes it out as a secret for general consumption.

### CODE REVIEW

File: bin/secrets [hunk 1]
```diff
@@ -6,13 +6,20 @@ vault_auth_pem = ENV["VAULT_AUTH_PEM"] || '/vault-auth/pem'
 vault_tls_verify = ['1', 'true'].include?(ENV["VAULT_TLS_VERIFY"])
 sidecar_secret_path = ENV["SIDECAR_SECRET_PATH"] || '/secrets'
 annotations = ENV["SECRET_ANNOTATIONS"] || '/secretkeys/annotations'
+serviceaccount_dir = ENV["SERVICEACCOUNT_DIR"] || '/var/run/secrets/kubernetes.io/serviceaccount/'
+api_host = ENV.fetch("KUBERNETES_PORT_443_TCP_ADDR")
```
Comment (grosser) [inline bin/secrets]: cannot use KUBERNETES_HOST or KUBERNETES_URL ?

File: lib/secrets.rb [hunk 2]
```diff
@@ -44,6 +48,8 @@ def write_secrets
       File.write("#{@output_path}/#{key}", contents)
       puts "Writing #{key} with contents from secret key #{path}"
     end
+    # Write out the pod's status.hostIP as a secret
+    File.write("#{@output_path}/HOST_IP", hostip)
```
Comment (grosser) [inline lib/secrets.rb]: would NODE_IP be clear ?

File: lib/secrets.rb [hunk 3]
```diff
@@ -65,6 +71,25 @@ def http_post(url, ssl_verify:, pem:)
     response.body
   end
 
+  def hostip
```
Comment (grosser) [inline lib/secrets.rb]: `def host_ip`

File: lib/secrets.rb [hunk 4]
```diff
@@ -65,6 +71,25 @@ def http_post(url, ssl_verify:, pem:)
...
+    token = File.read(@serviceaccount_dir + '/token')
+    namespace = File.read(@serviceaccount_dir + '/namespace')
+    uri = URI.parse(@api_url + "/api/v1/namespaces/#{namespace}/pods")
+    req = Net::HTTP::Get.new(uri.path)
+    req.add_field("Authorization", "Bearer #{token}")
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = (uri.scheme == 'https')
```
Comment (grosser) [inline lib/secrets.rb]: extract into a `http_get` method like the http_post method so this funky logic can stay in it's own little corner

File: lib/secrets.rb [hunk 5]
```diff
@@ -65,6 +71,25 @@ def http_post(url, ssl_verify:, pem:)
...
+    http.ca_file = "#{@serviceaccount_dir}/ca.crt"
+    puts uri.host
```
Comment (grosser) [inline lib/secrets.rb]: puts can go away ?

File: test/secrets_test.rb [hunk 6]
```diff
@@ -90,5 +99,16 @@ def process
...
     end
+
+    describe "api failures" do
+      before do
+        stub_request(:get, "https://foo.bar/api/v1/namespaces/default/pods").
+          to_return(status: 500)
+      end
+
+      it "raises when api calls fail" do
+        assert_raises(RuntimeError) { process }
```
Comment (grosser) [inline test/secrets_test.rb]: maybe a `e.message.must_include "Something"` so we can separate it from regular errors

File: test/secrets_test.rb [hunk 7]
```diff
@@ -19,20 +19,29 @@ def process
...
   end
   let(:auth_reply) { {auth: {client_token: 'sometoken'}}.to_json }
+  let(:status_api_body) { {items: [{status: {hostIP: "10.10.10.10"}}]}.to_json }
 
   before do
     stub_request(:post, "https://foo.bar:8200/v1/auth/cert/login").
       to_return(body: auth_reply)
+    stub_request(:get, "https://foo.bar/api/v1/namespaces/default/pods").
+      to_return(body: status_api_body)
+    ENV["KUBERNETES_PORT_443_TCP_ADDR"] = 'foo.bar'
```
Comment (grosser) [inline test/secrets_test.rb]: use with_env / do not modify global env state without resetting it ... or set it at the top of the test file

**Comment by grosser** [comment]
I like the random useful info ... pretty sure we will need more of this later on :)

File: bin/secrets (hunk 1 above)
Comment (irwaters) [inline bin/secrets]: nope, the options are:
KUBERNETES_PORT=tcp://10.247.0.1:443
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_SERVICE_PORT=443
KUBERNETES_SERVICE_HOST=10.247.0.1
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT_443_TCP_ADDR=10.247.0.1
KUBERNETES_PORT_443_TCP=tcp://10.247.0.1:443

so I could switch it to KUBERNETES_SERVICE_HOST if you'd prefer.

File: lib/secrets.rb (hunk 2 above)
Comment (irwaters) [inline lib/secrets.rb]: it's called hostIP in the pod definition, trying to stay close to what they have there.

File: test/secrets_test.rb [hunk 8]
```diff
@@ -90,5 +99,17 @@ def process
       reply.replace({foo: {bar: 1}}.to_json)
       assert_raises(RuntimeError) { process }
     end
+
```
Comment (grosser) [inline test/secrets_test.rb]: `it 'creates a HOST_IP secret'` test ?

File: lib/secrets.rb (hunk 4 above)
Comment (grosser) [inline lib/secrets.rb]: I was hoping for:

```Ruby
def http_get(url, headers: {}, ca_cert: nil)
```

so the 'http get logic' is separated from the 'vault request logic'

**Comment by grosser** [comment]
👍

**Pull request merged**